crates/jrsonnet-evaluator/src/evaluate/mod.rs

```diff
@@ -411,235 +413,852 @@ pub fn evaluate_named(ctx: Context, expr: &LocExpr, name: IStr) -> Result<Val> {
+	{
+		Tag {
+			_marker: PhantomData,
+		}
+	}
+}
+#[inline(always)]
+pub(crate) fn ctx_tag(name: &'static str) -> Tag<Context> {
+	#[cfg(debug_assertions)]
+	{
+		static ID: core::sync::atomic::AtomicU64 = core::sync::atomic::AtomicU64::new(0);
+		Tag {
+			name,
+			id: ID.fetch_add(1, core::sync::atomic::Ordering::SeqCst),
+			_marker: PhantomData,
+		}
+	}
+	#[cfg(not(debug_assertions))]
+	{
+		Tag {
+			_marker: PhantomData,
+		}
+	}
+}
+
+#[inline(always)]
+pub(crate) fn apply_tag<'a>() -> Tag<TailCallApply> {
+	#[cfg(debug_assertions)]
+	{
+		Tag {
+			name: "APPLY",
+			id: 0,
+			_marker: PhantomData,
 		}
-		Index(value, index) => match (evaluate(ctx.clone(), value)?, evaluate(ctx, index)?) {
-			(Val::Obj(v), Val::Str(key)) => State::push(
-				CallLocation::new(loc),
-				|| format!("field <{key}> access"),
-				|| match v.get(key.clone().into_flat()) {
-					Ok(Some(v)) => Ok(v),
-					Ok(None) => {
-						let suggestions = suggest_object_fields(&v, key.clone().into_flat());
-
-						throw!(NoSuchField(key.clone().into_flat(), suggestions))
+	}
+	#[cfg(not(debug_assertions))]
+	{
+		Tag {
+			_marker: PhantomData,
+		}
+	}
+}
+
+#[derive(Debug)]
+pub(crate) enum TailCallApply {
+	Eval {
+		expr: LocExpr,
+		in_ctx: Tag<Context>,
+		out_val: Tag<Val>,
+	},
+	ParseNormalArgs {
+		params: ParamsDesc,
+		args: ArgsDesc,
+		tailstrict: bool,
+		def_ctx: Tag<Context>,
+		call_ctx: Tag<Context>,
+		out_ctx: Tag<Context>,
+	},
+	IfCond {
+		in_val: Tag<Val>,
+		then_else_ctx: Tag<Context>,
+		then_val: LocExpr,
+		else_val: Option<LocExpr>,
+		out_val: Tag<Val>,
+	},
+	ApplyUnknownFn {
+		args: ArgsDesc,
+		location: CallLocation<'static>,
+		tailstrict: bool,
+		lhs_val: Tag<Val>,
+		call_ctx: Tag<Context>,
+		out_val: Tag<Val>,
+	},
+	ApplyBopSpecial {
+		and: bool,
+		a_val: Tag<Val>,
+		b_ctx: Tag<Context>,
+		b_expr: LocExpr,
+		out_val: Tag<Val>,
+	},
+	ApplyBopNormal {
+		op: BinaryOpType,
+		a_val: Tag<Val>,
+		b_val: Tag<Val>,
+		out_val: Tag<Val>,
+	},
+	ApplyUop {
+		op: UnaryOpType,
+		in_val: Tag<Val>,
+		out_val: Tag<Val>,
+	},
+	EvalObj {
+		in_ctx: Tag<Context>,
+		obj: RcMap<Expr, ObjBody>,
+		out_val: Tag<Val>,
+	},
+	AssertType {
+		in_val: Tag<Val>,
+		out_val: Tag<Val>,
+		ty: ValType,
+	},
+	EvalIndex {
+		lhs_val: Tag<Val>,
+		rhs_val: Tag<Val>,
+		out_val: Tag<Val>,
+	},
+	Unbound {
+		invoke: MaybeUnbound,
+		sup: Option<ObjValue>,
+		this: Option<ObjValue>,
+		out_val: Tag<Val>,
+	},
+}
+#[derive(Debug)]
+pub(crate) struct RcMap<T: ?Sized, U: ?Sized>(Rc<T>, fn(&T) -> &U);
+
+impl<T: ?Sized, U: ?Sized> RcMap<T, U> {
+	fn map(rc: Rc<T>, projection: fn(&T) -> &U) -> Self {
+		RcMap(rc, projection)
+	}
+}
+
+impl<T, U> Deref for RcMap<T, U>
+where
+	T: ?Sized,
+	U: ?Sized,
+{
+	type Target = U;
+
+	fn deref(&self) -> &Self::Target {
+		(self.1)(&*self.0)
+	}
+}
+
+pub(crate) struct TcVM {
+	pub(crate) vals: Fifo<Val>,
+	pub(crate) ctxs: Fifo<Context>,
+	pub(crate) apply: Fifo<TailCallApply>,
```

Comment:
Right here I have 3 stacks, where TailCallApply is an enum with items of varying length, that contains references to other stacks.
This enum value is pretty big, with no means to shrink without moving some things to allocation.
rsjsonnet did funny thing here, where it has 9 stacks instead, and keeps allocations even, because all the variable-size data is in the other stacks. This is pretty verbose code, with some implicit relations...

But I think it is possible to extend this approach using proc macro:
```rust
#[derive(FifoData)]
enum Op {
  A { a: u32, b: u32 },
  B { v: String },
  C { a: Context, b: Val, c: u32, d: String }
  // There should be a way to opt-out from moving data to stacks,
  // the remaining data might be attached to... OpTag enum?
}
```

Which would expand into
```rust
enum OpTag {
  A,
  B,
  C,
}

struct OpFifo {
  ops: Vec<OpTag>,
  u32s: Vec<u32>,
  strings: Vec<String>,
  // Where Val can also be inlined to some extent together with its children, so this proc macro
  // should allow composition of its elements
  vals: Vec<Val>,
  contexts: Vec<Context>,
}

impl Op {
  fn pop(fifo: &mut OpFifo) -> Self {
    match fifo.ops.pop() {
      OpTag::A => Self::A { b: fifo.u32s.pop(), a: fifo.u32s.pop() },
      OpTag::B => Self::B { v: fifo.strings.pop() },
      OpTag::C => Self::C { d: fifo.strings.pop(), c: fifo.u32s.pop(), b: fifo.vals.pop() },
    }
  }
  fn push(self, fifo: &mut OpFifo) {
    match self {
      Self::A {a, b} => { fifo.ops.push(OpTag::A); fifo.u32s.push(a); fifo.u32s.push(b); },
      _ => todo!(),
    }
  }
}
```

This structure is kinda dumb for the provided enum, but might make sense for the whole jrsonnet?